### DISCUSSION TITLE
RS422 input?

### DISCUSSION THREAD

**Discussion by wjquigs11** [hatlabs/discussions#61] [Q&A]
Edit: @mairas mentioned an RS422 module in a different thread...what's the function of that device if the ESP32 has several serial interfaces already? Possibly because they are digital interfaces and the AIS output is analog?

(Also I understand I will need to use Serial1 or 2 so as not to conflict with the usb port on Serial0)

Has anyone used UART0 on the Sailor HAT to read NMEA 0183 data?
Presumably I can initialize using something like this:

#define RXD0 03
#define TXD0 01
Serial0.begin(9600, SERIAL_8N1, RXD0, TXD0);

If I have NMEA 0183 output (RS422) from my AIS configured correctly and connected to GPIO pin 03, hopefully I can read from Serial0 and can interpret using the library from ttlappalainen. (Initially just looking for GPS fix; not sure if his library includes AIS-specific PGNs but there are others available).

Is it as simple as this?

**Comment by nousian** [comment]
> the AIS output is analog

What do you mean by this? It is usually either 0183 or 2000, all digital, I think.

Anyways....

EspSoftwareSerial should work fine with the optocoupler pins on the four pin header - that is what I am using right now, though mine is connected to Actisense NDC-3 multiplexer RS232 output.

Relevant snippets from my code:

```
#include <SoftwareSerial.h>

#define MYPORT_TX GPIO_NUM_33
#define MYPORT_RX GPIO_NUM_35

EspSoftwareSerial::UART myPort;

void handleSerial() {
  serialReads++; // counters I added

  while (myPort.available() > 0) {
    serialBytes++; // counters I added
    uint8_t c = myPort.read();

  // -- process the chars read...
  }
}

void setupSoftSerial() {
  myPort.begin(4800, SWSERIAL_8N1, MYPORT_RX, MYPORT_TX, true);

  if (myPort)
    app.onRepeat(10, []() { handleSerial(); });
}

void setup() {
  setupSoftSerial();
  // other stuff
}
```
